aten/src/ATen/native/TensorShape.cpp

```diff
@@ -2149,33 +2149,83 @@ std::vector<Tensor> meshgrid(TensorList tensors,
   int64_t size = tensors.size();
   TORCH_CHECK(size > 0, "meshgrid expects a non-empty TensorList");
 
-  TORCH_CHECK(
-      indexing == "ij",
-      "torch.meshgrid: only \"ij\" indexing is supported at this time, but "
-      "received: ", indexing);
+  for(const auto i: c10::irange(size - 1)){
+    TORCH_CHECK(tensors[i].dtype() == tensors[i+1].dtype(), "meshgrid expects all tensors to have the same dtype");
+    TORCH_CHECK(tensors[i].device() == tensors[i+1].device(), "meshgrid expects all tensors to have the same device");
```

Comment:
This check is auto-generated as long as the op doesn't opt-out of it in native_functions.yaml

Reply:
interestingly enough: this does not work for this function, despite not having opted out of it.

```
>>> torch.meshgrid(torch.rand(1, device='cpu'), torch.rand(1, device='cuda'))
(tensor([[0.8120]]), tensor([[0.9082]], device='cuda:0'))
```

Does this feature not support TensorList arguments? Is that an oversight?

Reply:
Huh; interesting. Maybe it doesn't support lists -- good catch

Reply:
Do we want to add list support to that feature?

Reply:
I'm not sure what the impact of doing so would be, but you could file an issue. I'm also not sure if anyone is actively maintaining that code, though

Reply:
So `Tensor[]` is supported, but device check is only performed for CUDA kernels.

So the questions are:
1) should device check also apply to non-cuda implementations?
2) does meshgrid need a device check? I think it works efficiently even if the inputs are across devices, since each output corresponds to a single tensor input. e.g. if you pass a cpu and a cuda tensor, you get back a cpu and a cuda tensor.

Reply:
From offline conversation:

- We should at least clarify when the check applies
- We should create a simple rule for all ops like "composites should always implement their own check" but "prims can rely on the generated check"